src/entities/addLiquidityBoosted/index.ts

```diff
@@ -0,0 +1,187 @@
+import { doAddLiquidityProportionalQuery } from './doAddLiquidityPropotionalQuery';
+import { Token } from '../token';
+import { BALANCER_COMPOSITE_LIQUIDITY_ROUTER } from '@/utils';
+
+import { getValue } from '@/entities/utils/getValue';
+import { encodeFunctionData } from 'viem';
+import { balancerCompositeLiquidityRouterAbi, balancerRouterAbi } from '@/abi';
+
+import { Hex } from '@/types';
+
+export class AddLiquidityBoostedV3 {
+    async query(
+        input: AddLiquidityInput,
+        poolState: PoolState | PoolStateWithUnderlyings,
+    ): Promise<AddLiquidityBaseQueryOutput> {
+        // Technically possible to pass singleToken adds here due to type
+        // disallow it for this class
+        // TODO: Use input validator
+        const bptToken = new Token(input.chainId, poolState.address, 18);
+
+        let bptOut: TokenAmount;
+        let amountsIn: TokenAmount[];
+        let tokenInIndex: number | undefined;
+        switch (input.kind) {
+            case AddLiquidityKind.Unbalanced: {
+                const bptAmountOut = await doAddLiquidityUnbalancedQuery(
+                    input,
+                    poolState.address,
+                );
+                bptOut = TokenAmount.fromRawAmount(bptToken, bptAmountOut);
+
+                amountsIn = input.amountsIn.map((t) => {
+                    return TokenAmount.fromRawAmount(
+                        new Token(input.chainId, t.address, t.decimals),
+                        t.rawAmount,
+                    );
+                });
+
+                break;
+            }
+            case AddLiquidityKind.Proportional: {
+                const exactAmountsInNumbers =
+                    await doAddLiquidityProportionalQuery(
+                        input,
+                        poolState.address,
+                    );
+
+                // Since the user adds tokens which are technically not pool tokens, the TokenAmount to return
+                // uses the pool's tokens underlyingTokens to indicate which tokens are being added from the user
+                // perspective
+                amountsIn = poolState.tokens.map((t, i) =>
+                    TokenAmount.fromRawAmount(
+                        new Token(
+                            input.chainId,
+                            t.underlyingToken.address,
+                            t.underlyingToken.decimals,
+                            t.underlyingToken.symbol,
+                        ),
+                        exactAmountsInNumbers[i],
+                    ),
+                );
+
+                bptOut = TokenAmount.fromRawAmount(
+                    bptToken,
+                    input.referenceAmount.rawAmount,
+                );
+                break;
+            }
+            case AddLiquidityKind.SingleToken: {
+                throw new Error('SingleToken not supported');
+            }
+        }
+
+        const output: AddLiquidityBaseQueryOutput = {
```

Comment:
In the nested PR I update to return userAddress and userData to buildCall can use that info.

Reply:
I have addressed this for add- as well as remove liquidity